Source/api/Stats/Manager/StatsValueDocument.cs

```
@@ -122,23 +154,71 @@ internal List<string> GetStatNames()
+                                    }
+                                    else
+                                    {
+                                        Stats[svdEvent.Name].SetStat(svdEvent.Value, svdEvent.Type);
+                                    }
+
+                                    IsDirty = true;
+                                    break;
+                                }
+
+                            case StatPendingEventType.StatDelete:
+                                {
+                                    Stats.Remove(svdEvent.Name);
+                                    break;
+                                }
+                        }
                     }
-                    else
+
+                    this.pendingEventList.Clear();
+                }
+            }
+        }
+
+        internal void MergeStatDocument(StatsValueDocument mergeStatDocument)
+        {
+            switch (State)
+            {
+                case StatValueDocumentState.NotLoaded:
+                    Revision = mergeStatDocument.Revision;
+                    Stats = mergeStatDocument.Stats;
+                    break;
+
+                // for offline the stat values local override any service values
+                // only add any undefined stats into our list
+                case StatValueDocumentState.OfflineNotLoaded:
+                case StatValueDocumentState.OfflineLoaded:
+                    foreach (var stat in mergeStatDocument.Stats)
                     {
-                        Stats[svdEvent.Name].SetStat(svdEvent.Value, svdEvent.Type);
+                        if (Stats.ContainsKey(stat.Key))
+                        {
+                            Stats.Add(stat.Key, stat.Value);
+                        }
                     }
-                }
+                    break;
 
-                this.eventList.Clear();
+                case StatValueDocumentState.Loaded:
```

This should probably throw an exception because I don't think we should be getting here multiple times simultaneously.